File: frontend/src/api/updates.ts
import { api } from './client'

export interface UpdateStatus {
  current_version: string
  latest_version: string | null
  update_available: boolean
  notes: string | null
  published_at: string | null
  error: string | null
  publish_enabled: boolean
}

export interface ApplyUpdateResult {
  ok: boolean
  version: string
  log: string
  restarting: boolean
  restart_mode: 'supervised' | 'relaunch' | 'manual'
}

export interface PublishUpdateResult {
  html_url: string
  tag_name: string
  asset_name: string
  asset_size: number
}

export interface UpdateSource {
  kind: 'server' | 'github'
  address: string
  repo: string
  token_set: boolean
}

export interface AvailableUpdate {
  version: string
  notes: string
  published_at: string
  size: number | null
  current: boolean
  newer: boolean
}

export interface ChangelogOut {
  version: string
  markdown: string
}

export interface UpdateAttempt {
  at: string
  from_version: string
  to_version: string
  ok: boolean
  detail: string
}

export interface RestorePoint {
  name: string
  version: string
  at: string
}

export const updatesApi = {
  /** What this PC has actually been through, newest first - the failures too. */
  history: () => api.get<UpdateAttempt[]>('/updates/history'),
  restorePoints: () => api.get<RestorePoint[]>('/updates/restore-points'),
  /** The release notes for the version this PC is actually running - read
   *  from the CHANGELOG.md that shipped in the package with it. */
  changelog: () => api.get<ChangelogOut>('/updates/changelog'),
  status: () => api.get<UpdateStatus>('/updates/status'),
  // Skips the server's 15-minute cache - the "Check now" button.
  checkNow: () => api.get<UpdateStatus>('/updates/status?force=true'),
  apply: () => api.post<ApplyUpdateResult>('/updates/apply', {}),
  applyVersion: (version: string, allowOlder: boolean) =>
    api.post<ApplyUpdateResult>('/updates/apply', { version, allow_older: allowOlder }),

  source: () => api.get<UpdateSource>('/updates/source'),
  setSource: (address: string, token: string | null) =>
    api.put<UpdateSource>('/updates/source', { address, token }),
  available: () => api.get<AvailableUpdate[]>('/updates/available'),
  publish: (from_version: string, notes: string) =>
    api.post<PublishUpdateResult>('/updates/publish', { from_version, notes }),
}

File: api/schemas/updates.py
from pydantic import BaseModel


class UpdateStatusOut(BaseModel):
    current_version: str
    latest_version: str | None
    update_available: bool
    notes: str | None = None
    published_at: str | None = None
    error: str | None = None
    # Whether THIS machine can build and sign a release (MES_ALLOW_PUBLISH=1
    # in its .env) - a plant PC has neither the key nor the dev tools, so the
    # page hides that half rather than offering a button that can only fail.
    publish_enabled: bool = False


class ApplyUpdateOut(BaseModel):
    ok: bool
    version: str
    log: str
    restarting: bool
    # "supervised" (the launcher will start it again), "relaunch" (nothing is
    # watching, so a helper will), or "manual" (somebody has to).
    restart_mode: str = "manual"


class PublishUpdateRequest(BaseModel):
    from_version: str = ""
    notes: str = ""


class PublishUpdateOut(BaseModel):
    html_url: str
    tag_name: str
    asset_name: str
    asset_size: int


class UpdateSourceOut(BaseModel):
    kind: str          # "server" (one you host) or "github"
    address: str       # the update server's address, when that's the source
    repo: str          # the GitHub repo, when that is
    token_set: bool


class UpdateSourceRequest(BaseModel):
    address: str = ""          # blank goes back to GitHub
    token: str | None = None   # None leaves whatever is already set alone


class UpdateAttempt(BaseModel):
    """One line out of updates/history.jsonl - what this PC actually went
    through, successes and failures alike."""
    at: str = ""
    from_version: str = ""
    to_version: str = ""
    ok: bool = False
    detail: str = ""


class RestorePoint(BaseModel):
    """A copy-aside taken before an update, still sitting in rollback\."""
    name: str
    version: str = ""
    at: str = ""


class AvailableUpdate(BaseModel):
    version: str
    notes: str = ""
    published_at: str = ""
    size: int | None = None
    current: bool = False
    newer: bool = False


class ApplyVersionRequest(BaseModel):
    version: str = ""          # blank means "the newest one"
    allow_older: bool = False  # installing an older release, deliberately
